File: 0147_insertion_sort_list/insertion_sort.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def insertionSortList(head: ListNode) -> ListNode:
    """
    Time: O(n**2)
    Space: O(1)
    """
    s = ListNode()
    while head:
        left, right = s, s.next
        while right:
            if head.val < right.val:
                break
            left, right = right, right.next
        _next = head.next
        left.next = head
        head.next = right
        head = _next
    return s.next
        

File: 0147_insertion_sort_list/insertion_sort.java
public class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    // Time: O(n**2)
    // Space: O(1)
    public ListNode insertionSortList(ListNode head) {
        ListNode sentinel = new ListNode();
        ListNode curr = head, left, right, next;
        while (curr != null) {
            left = sentinel;
            right = sentinel.next;
            while (right != null) {
                if (curr.val < right.val) {
                    break;
                }
                left = right;
                right = right.next;
            }
            next = curr.next;
            left.next = curr;
            curr.next = right;
            curr = next;
        }
        return sentinel.next;       
    }
}